### PR TITLE
Additional data for public alarm ack API

### LINKED ISSUE #147: Segmentation fault in pf_alarm.c
File **pf_alarm.c**:

The function `pf_alarm_apms_a_data_req()` has as its fourth parameter `pf_alarm_data_t  *p_alarm_data`, which is then used to store alarm data by `pf_put_alarm_block()` function. This parameter is never checked against NULL.
But the `pf_alarm_apms_a_data_req()` is called with its **p_alarm_data** parameter set to NULL in the `pf_alarm_alpmr_alarm_ack()` function. 
This subsequently leads to segmentation fault (i.e. dereferencing NULL pointer in the `pf_put_alarm_block()` function).

Somewhere a check against NULL is necessary, either in the `pf_alarm_apms_a_data_req()` or maybe in the low-level function `pf_put_alarm_block()`.

### PR DESCRIPTION
**Pull request by pyhys**
Improved transport ACK (TACK) handling.
Corrected writing of outgoing messages.
Closes #147

### CODE REVIEW

File: sample_app/main_rtk.c [hunk 1]
```diff
@@ -286,7 +286,11 @@ int main(void)
 
          os_event_clr(appdata.main_events, EVENT_ALARM); /* Re-arm */
 
-         ret = pnet_alarm_send_ack(g_net, appdata.main_arep, &pnio_status);
+         ret = pnet_alarm_send_ack(g_net, appdata.main_arep,
```
Comment (hefloryd) [inline sample_app/main_rtk.c:289]: That's quite a lot of arguments. Would it make sense to introduce an alarm type (consisting of slot, subslot, type, sequence_number and specifier)?

Comment (pyhys) [reply sample_app/main_rtk.c:289]: Yes, I can do that.

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]
Low-level review only, as I'm not familiar with the functionality

File: sample_app/main_rtk.c (hunk 1 above)
Comment (pyhys) [reply sample_app/main_rtk.c]: Implemented an alarm argument type

File: include/pnet_api.h [hunk 2]
```diff
@@ -764,6 +781,9 @@ typedef int (*pnet_alarm_ind)(
    uint32_t                api,
    uint16_t                slot,
    uint16_t                subslot,
+   uint16_t                alarm_type,
+   uint16_t                alarm_sequence_number,
+   pnet_alarm_spec_t       *p_alarm_specifier,
```
[the remark this answers is not in the document]
Comment (pyhys) [reply include/pnet_api.h]: Replaced with
`const pnet_alarm_argument_t   *p_alarm_argument,`

File: src/device/pf_block_writer.h [hunk 3]
```diff
@@ -506,6 +504,7 @@ void pf_put_alarm_block(
    uint16_t                payload_usi,
    uint16_t                payload_len,
    uint8_t                 *p_payload,
+   pnet_pnio_status_t      *p_status,
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_block_writer.h:507]: Done

File: src/device/pf_fspm.h [hunk 4]
```diff
@@ -204,6 +210,9 @@ int pf_fspm_alpmr_alarm_ind(
    uint32_t                api,
    uint16_t                slot,
    uint16_t                subslot,
+   uint16_t                alarm_type,
+   uint16_t                alarm_sequence_number,
+   pnet_alarm_spec_t       *p_alarm_specifier,
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_fspm.h]: Replaced with
`const pnet_alarm_argument_t   *p_alarm_arg,`

File: src/common/pf_alarm.c [hunk 5]
```diff
@@ -53,6 +53,28 @@ static const char *apmx_sync_name = "apmx";
...
+/**
+ * @internal
+ * Return a string representation of alarm PDU type (message type).
+ * @param pdu_type            In:   alarm PDU type  See pf_alarm_pdu_type_values_t
+ * @return  A string representing the alarm PDU type.
+ */
+static const char *pf_alarm_pdu_type_to_string(
+   uint8_t pdu_type)
+{
+   const char *s = "<error>";
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/common/pf_alarm.c:65]: Done

File: src/common/pf_alarm.h [hunk 6]
```diff
@@ -202,17 +202,34 @@ int pf_alarm_send_diagnosis(
...
- * Send Alarm ACK.
+ * Send Alarm ACK (from the application)
  *
- * @param net              InOut: The p-net stack instance
- * @param p_ar             In:   The AR instance.
- * @param p_pnio_status    In:   Detailed ACK information.
+ * Always uses high prio.
+ *
+ * ALPMR: ALPMR_Alarm_Ack.req
+ * ALPMR: ALPMR_Alarm_Ack.cnf   part of   via return value
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/common/pf_alarm.h]: Changed to 
`ALPMR: ALPMR_Alarm_Ack.cnf   (Implements part of that signal via the return value)`

**Review by fredrikdanielmoller — APPROVED** [review]

**Pull request merged**